### PR TITLE
Station alert levels

### PR DESCRIPTION
**Pull request by vulppine**
## About the PR 

Adds alert levels to stations. Alert levels are definable via a prototype, `alertLevels`, which has a dictionary of alert level IDs to alert level definitions. Currently, stations can only have one alert level set due to how StationSystem works, but otherwise, alerts can be added or removed without having to delve into C#.

:cl:
- add: NanoTrasen has installed a new alert readiness system into its stations! Use these to indicate to your fellow crewmembers when the station is going OK, or is in a flurry of viral infections, hostile activity, and imminent destruction!

### CODE REVIEW

File: Content.Server/Atmos/Monitor/Systems/FireAlarmSystem.cs [hunk 1]
```diff
@@ -16,6 +18,18 @@ public override void Initialize()
         {
             SubscribeLocalEvent<FireAlarmComponent, InteractHandEvent>(OnInteractHand);
             SubscribeLocalEvent<FireAlarmComponent, GotEmaggedEvent>(OnEmagged);
+            SubscribeLocalEvent<AlertLevelChangedEvent>(OnAlertLevelChanged);
+        }
+
+        private void OnAlertLevelChanged(AlertLevelChangedEvent args)
+        {
+            foreach (var entity in EntityManager.EntityQuery<FireAlarmComponent>())
```
Comment (metalgearsloth) [inline Content.Server/Atmos/Monitor/Systems/FireAlarmSystem.cs:26]: EntityQuery uses generics so you can get back multiple components.

File: Content.Server/AlertLevel/AlertLevelPrototype.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    /// <summary>
+    /// The sound that this alert level will play in-game once selected.
+    /// </summary>
+    [DataField("sound")] public SoundSpecifier? Sound { get; }
+
+    /// <summary>
+    /// The color that this alert level will show in-game in chat.
+    /// </summary>
+    [DataField("color")] public string Color { get; } = default!;
```
Comment (metalgearsloth) [inline Content.Server/AlertLevel/AlertLevelPrototype.cs:59]: Should just be able to make this a color.

File: Content.Server/AlertLevel/AlertLevelPrototype.cs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+/// Alert level detail. Does not contain an ID, that is handled by
+/// the Levels field in AlertLevelPrototype.
+/// </summary>
+[DataDefinition]
+public sealed class AlertLevelDetail
+{
+    /// <summary>
+    /// What is announced upon this alert level change. Can be a localized string.
+    /// </summary>
+    [DataField("announcement")] public string Announcement { get; } = default!;
```
Comment (metalgearsloth) [inline Content.Server/AlertLevel/AlertLevelPrototype.cs:36]: nullable string?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Just a quick review, need to suss out ingame.

**Comment by Peptide90** [comment]
Sweet! Noticed the visualiser is on the fire alarm though? Is the alert status screen for someone else to add or is this intended?

File: Content.Server/AlertLevel/AlertLevelPrototype.cs (hunk 3 above)
Comment (vulppine) [reply Content.Server/AlertLevel/AlertLevelPrototype.cs:36]: nah, it should still be at the very least an empty string if there's no real announcement - maybe `default!` isn't the best thing here and string.Empty should be the better replacement?

File: Content.Client/AlertLevel/AlertLevelDisplayVisualizer.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+        }
+
+        var layer = sprite.AddLayer(new RSI.StateId(_alertVisuals.Values.First()));
+        sprite.LayerMapSet(AlertLevelDisplay.Layer, layer);
+    }
+
+    public override void OnChangeData(AppearanceComponent component)
+    {
+        if (!component.TryGetData<string>(AlertLevelDisplay.CurrentLevel, out var level)
+            || !_entityManager.TryGetComponent(component.Owner, out SpriteComponent? sprite))
```
[the remark this answers is not in the document]
Comment (vulppine) [reply Content.Client/AlertLevel/AlertLevelDisplayVisualizer.cs]: the appearance change event has AppearanceComponent, but not SpriteComponent

**Comment by vulppine** [comment]
addressed reviews

**Comment by vulppine** [comment]
> Sweet! Noticed the visualiser is on the fire alarm though? Is the alert status screen for someone else to add or is this intended?

iirc the status screen had more than just alerts, right? in any case, i was able to make it so that AlertLevelDisplay now can just be added to any entity without any issue (as long as the layers look nice on the entity, at least)

File: Content.Client/AlertLevel/AlertLevelDisplayVisualizer.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (metalgearsloth) [reply Content.Client/AlertLevel/AlertLevelDisplayVisualizer.cs]: If you merge to master it should be there, it was changed 2 days ago.

File: Content.Server/AlertLevel/AlertLevelPrototype.cs (hunk 3 above)
Comment (metalgearsloth) [reply Content.Server/AlertLevel/AlertLevelPrototype.cs:36]: yeah I'd just go with a string.Empty and make it a required field then.

**Review by moonheart08 — APPROVED** [review]

**Comment by metalgearsloth** [comment]
I made some comments in dms in case a maintainer was gonna merge:
1. Blue is too dark
2. You can re-select the same alert level
3. Alert level should have some kinda description rather than colour (e.g. "Yellow (atmospheric threat)")

File: Content.Server/AlertLevel/AlertLevelPrototype.cs (hunk 3 above)
Comment (vulppine) [reply Content.Server/AlertLevel/AlertLevelPrototype.cs]: it's theoretically fine to have an empty announcement, but it's just lazy

**Comment by vulppine** [comment]
that should be everything i want to do / all reviews addressed sans the alert description, should be good to merge

**Pull request merged**